### PR TITLE
`azurerm_log_analytics_solution`: fix update & refactor to typed sdk

### LINKED ISSUE #23312: azurerm_log_analytics_solution unexpected status 200 with response: {"plan":{"name": . . .
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.5.1

### AzureRM Provider Version

3.73.0.

### Affected Resource(s)/Data Source(s)

azurerm_log_analytics_solution

### Terraform Configuration Files

```hcl
# Not possible to reproduce with minimal config


resource "azurerm_log_analytics_solution" "solution" {
  for_each = var.solution_names

  solution_name         = each.key
  location              = var.location
  resource_group_name   = var.resource_group_name
  workspace_resource_id = var.workspace_resource_id
  workspace_name        = var.workspace_name

  plan {
    publisher = var.publisher
    product   = "OMSGallery/${each.key}"
  }

  tags = var.tags
}
```

### Debug Output/Panic Output

```shell
╷
│ Error: creating/updating Solution (Subscription: "***"
│ Resource Group Name: "rg-sdp-prod"
│ Solution Name: "ContainerInsights(log-prod)"): performing CreateOrUpdate: unexpected status 200 with response: {"plan":{"name":"ContainerInsights(log-prod)","publisher":"Microsoft","promotionCode":"","product":"OMSGallery/ContainerInsights"},"properties":{"workspaceResourceId":"/subscriptions/***/resourceGroups/rg-sdp-prod/providers/Microsoft.OperationalInsights/workspaces/log-prod","provisioningState":"Succeeded","creationTime":"Thu, 11 May 2023 10:11:26 GMT","lastModifiedTime":"Tue, 19 Sep 2023 07:50:54 GMT","containedResources":[],"workbookTemplates":[]},"location":"northeurope","tags":{"AppName":"ADAS - Advanced Driver Assistance System","Environment":"Production","StableforReservation":"No","WBS Code":"P19111"},"id":"/subscriptions/***/resourceGroups/rg-sdp-prod/providers/Microsoft.OperationsManagement/solutions/ContainerInsights(log-prod)","name":"ContainerInsights(log-prod)","type":"Microsoft.OperationsManagement/solutions"}
│ 
│   with module.loganalytics_solutions.azurerm_log_analytics_solution.solution["ContainerInsights"],
│   on .terraform/modules/loganalytics_solutions/loganalytics_solution/main.tf line 1, in resource "azurerm_log_analytics_solution" "solution":
│    1: resource "azurerm_log_analytics_solution" "solution" {
│ 
╵
Error: Process completed with exit code 1.
```

### Expected Behaviour

Just tags updated, as for PLAN below

```
# module.loganalytics_solutions.azurerm_log_analytics_solution.solution["ContainerInsights"] will be updated in-place
  ~ resource "azurerm_log_analytics_solution" "solution" {
        id                    = "/subscriptions/***/resourceGroups/rg-sdp-prod/providers/Microsoft.OperationsManagement/solutions/ContainerInsights(log-prod)"
      ~ tags                  = {
          - "AppName"              = "ADAS - Advanced Driver Assistance System" -> null
          - "Environment"          = "Production" -> null
          - "StableforReservation" = "No" -> null
          - "WBS Code"             = "P19111" -> null
        }
        # (5 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```

### Actual Behaviour

APPLY fails with error

```
╷
│ Error: creating/updating Solution (Subscription: "***"
│ Resource Group Name: "rg-sdp-prod"
│ Solution Name: "ContainerInsights(log-prod)"): performing CreateOrUpdate: unexpected status 200 with response: {"plan":{"name":"ContainerInsights(log-prod)","publisher":"Microsoft","promotionCode":"","product":"OMSGallery/ContainerInsights"},"properties":{"workspaceResourceId":"/subscriptions/***/resourceGroups/rg-sdp-prod/providers/Microsoft.OperationalInsights/workspaces/log-prod","provisioningState":"Succeeded","creationTime":"Thu, 11 May 2023 10:11:26 GMT","lastModifiedTime":"Tue, 19 Sep 2023 07:50:54 GMT","containedResources":[],"workbookTemplates":[]},"location":"northeurope","tags":{"AppName":"ADAS - Advanced Driver Assistance System","Environment":"Production","StableforReservation":"No","WBS Code":"P19111"},"id":"/subscriptions/***/resourceGroups/rg-sdp-prod/providers/Microsoft.OperationsManagement/solutions/ContainerInsights(log-prod)","name":"ContainerInsights(log-prod)","type":"Microsoft.OperationsManagement/solutions"}
│ 
│   with module.loganalytics_solutions.azurerm_log_analytics_solution.solution["ContainerInsights"],
│   on .terraform/modules/loganalytics_solutions/loganalytics_solution/main.tf line 1, in resource "azurerm_log_analytics_solution" "solution":
│    1: resource "azurerm_log_analytics_solution" "solution" {
│ 
╵
Error: Process completed with exit code 1.
```

### Steps to Reproduce

terraform plan
terraform apply

### Important Factoids

We provisioned in may with latest 3.x version of the provider available at that time. 
We had recently manually created a diagnostic setting in the AKS cluster using the workspace, but we have removed everything and restored the previous situation (all TF provisioned).

### References

No

### PR DESCRIPTION
**Pull request by ziyeqf**
fix #23312

from the memory, we should not slipt the `createUpdate` function into two functions, which will break resources' lifecycle, if I'm wrong please let me know and I will update the code.

test
---
```
❯ tftest loganalytics TestAccLogAnalyticsSolution                               
=== RUN   TestAccLogAnalyticsSolution_basicContainerMonitoring
=== PAUSE TestAccLogAnalyticsSolution_basicContainerMonitoring
=== RUN   TestAccLogAnalyticsSolution_update
=== PAUSE TestAccLogAnalyticsSolution_update
=== RUN   TestAccLogAnalyticsSolution_requiresImport
=== PAUSE TestAccLogAnalyticsSolution_requiresImport
=== RUN   TestAccLogAnalyticsSolution_basicSecurity
=== PAUSE TestAccLogAnalyticsSolution_basicSecurity
=== CONT  TestAccLogAnalyticsSolution_basicContainerMonitoring
=== CONT  TestAccLogAnalyticsSolution_requiresImport
=== CONT  TestAccLogAnalyticsSolution_basicSecurity
=== CONT  TestAccLogAnalyticsSolution_update
--- PASS: TestAccLogAnalyticsSolution_basicContainerMonitoring (222.62s)
--- PASS: TestAccLogAnalyticsSolution_requiresImport (228.58s)
--- PASS: TestAccLogAnalyticsSolution_basicSecurity (237.80s)
--- PASS: TestAccLogAnalyticsSolution_update (271.83s)
```

### CODE REVIEW

**Review by stephybun — CHANGES_REQUESTED** [review]
@ziyeqf even though this fixes the issue we are actively moving towards split Create/Update methods for all resources. Can you please split them completely instead of reverting to a legacy pattern or elaborate one what the issue with the lifecycle was?

**Comment by ziyeqf** [comment]
> @ziyeqf even though this fixes the issue we are actively moving towards split Create/Update methods for all resources. Can you please split them completely instead of reverting to a legacy pattern or elaborate one what the issue with the lifecycle was?

code updated. the test is here:
```
❯ tftest loganalytics TestAccLogAnalyticsSolution
=== RUN   TestAccLogAnalyticsSolution_basicContainerMonitoring
=== PAUSE TestAccLogAnalyticsSolution_basicContainerMonitoring
=== RUN   TestAccLogAnalyticsSolution_update
=== PAUSE TestAccLogAnalyticsSolution_update
=== RUN   TestAccLogAnalyticsSolution_requiresImport
=== PAUSE TestAccLogAnalyticsSolution_requiresImport
=== RUN   TestAccLogAnalyticsSolution_basicSecurity
=== PAUSE TestAccLogAnalyticsSolution_basicSecurity
=== CONT  TestAccLogAnalyticsSolution_basicContainerMonitoring
=== CONT  TestAccLogAnalyticsSolution_requiresImport
=== CONT  TestAccLogAnalyticsSolution_basicSecurity
=== CONT  TestAccLogAnalyticsSolution_update
--- PASS: TestAccLogAnalyticsSolution_basicSecurity (208.32s)
--- PASS: TestAccLogAnalyticsSolution_requiresImport (245.45s)
--- PASS: TestAccLogAnalyticsSolution_basicContainerMonitoring (261.45s)
--- PASS: TestAccLogAnalyticsSolution_update (301.09s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/loganalytics  301.151s
```

And, the "lifecycle" memory, I'm not sure where it come from..

File: internal/services/loganalytics/log_analytics_solution_resource.go [hunk 1]
```diff
@@ -4,255 +4,311 @@
...
+				return err
+			}
+
+			var config SolutionResourceModel
+			if err := metadata.Decode(&config); err != nil {
+				return fmt.Errorf("decoding %v", err)
+			}
 
-	return nil
+			payload := solution.SolutionPatch{
```
Comment (stephybun) [inline internal/services/loganalytics/log_analytics_solution_resource.go:289]: We should use `metadata.ResourceData.HasChange` on tags here.

**Review by stephybun — CHANGES_REQUESTED** [review]
@ziyeqf thanks for converting this over to a typed resource as well as splitting the Create/Update. One issue though is that we can't just remove state migrations since they become a permanent component of a resource. You can use the `StateUpgraders` method to add them for typed resources.

File: internal/services/loganalytics/log_analytics_solution_resource.go (hunk 1 above)
Comment (ziyeqf) [reply internal/services/loganalytics/log_analytics_solution_resource.go]: `tag` is the only property could be changed, I'm considering `HasChange()` will alwasy return `true`? or the `update` func will not be called.

Comment (stephybun) [reply internal/services/loganalytics/log_analytics_solution_resource.go]: We still need to call `HasChange` on the property for `ignore_changes` to work on it.

**Review by stephybun — APPROVED** [review]
Thanks @ziyeqf LGTM 👍

File: internal/services/loganalytics/log_analytics_solution_resource.go (hunk 1 above)
Comment (ziyeqf) [reply internal/services/loganalytics/log_analytics_solution_resource.go]: got it

**Pull request merged**